File: apps/api/app/models/assistant.py
"""Assistant conversation/message models (US-165, EP008).

Persistencia del asistente IA conversacional (widget global). Una
`AssistantConversation` agrupa el hilo de un usuario dentro de un tenant;
cada `AssistantMessage` es un turno (user/assistant) con las `actions`
estructuradas que el frontend puede ejecutar (navegar, sugerir, etc.).

A diferencia de `ai_jobs` (one-shot async), aquí el chat es sincrónico y
multi-turno; el historial vive en DB para sobrevivir reloads (ENH-076).
"""
from datetime import datetime
from uuid import UUID

from sqlalchemy import (
    JSON,
    DateTime,
    ForeignKey,
    Index,
    String,
    Text,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, new_uuid


class AssistantConversation(Base):
    __tablename__ = "assistant_conversations"
    __table_args__ = (
        Index("idx_assistant_conv_user", "tenant_id", "user_id", "updated_at"),
    )

    id: Mapped[UUID] = mapped_column(String(36), primary_key=True, default=new_uuid)
    tenant_id: Mapped[UUID] = mapped_column(
        String(36), ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False
    )
    user_id: Mapped[UUID] = mapped_column(
        String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False
    )
    title: Mapped[str | None] = mapped_column(String(200))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )


class AssistantMessage(Base):
    __tablename__ = "assistant_messages"
    __table_args__ = (
        Index("idx_assistant_msg_conv", "conversation_id", "created_at"),
    )

    id: Mapped[UUID] = mapped_column(String(36), primary_key=True, default=new_uuid)
    conversation_id: Mapped[UUID] = mapped_column(
        String(36),
        ForeignKey("assistant_conversations.id", ondelete="CASCADE"),
        nullable=False,
    )
    role: Mapped[str] = mapped_column(String(16), nullable=False)  # user | assistant
    content: Mapped[str] = mapped_column(Text, nullable=False, default="")
    # Acciones estructuradas que el frontend puede ejecutar (navegar, etc.).
    actions: Mapped[list] = mapped_column(JSON, nullable=False, default=list)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

File: apps/web/lib/api/assistant.ts
import { apiFetch } from "@/lib/api";

/** US-165 — cliente del asistente IA conversacional (widget global). */

export type AssistantAction = {
  type: "navigate" | "none";
  path?: string | null;
  label?: string | null;
};

export type AssistantChatResponse = {
  conversation_id: string;
  message: string;
  actions: AssistantAction[];
};

export type AssistantMessage = {
  id: string;
  role: "user" | "assistant";
  content: string;
  actions: AssistantAction[];
  created_at: string;
};

export type AssistantConversation = {
  id: string;
  title: string | null;
  created_at: string;
  updated_at: string;
};

export type AssistantConversationDetail = AssistantConversation & {
  messages: AssistantMessage[];
};

export function sendAssistantMessage(input: {
  message: string;
  conversation_id?: string | null;
  page_context?: string | null;
}): Promise<AssistantChatResponse> {
  return apiFetch<AssistantChatResponse>("/api/v1/assistant/chat", {
    method: "POST",
    body: {
      message: input.message,
      conversation_id: input.conversation_id ?? null,
      page_context: input.page_context ?? null,
    },
  });
}

export function listAssistantConversations(): Promise<AssistantConversation[]> {
  return apiFetch<AssistantConversation[]>("/api/v1/assistant/conversations");
}

export function getAssistantConversation(
  id: string,
): Promise<AssistantConversationDetail> {
  return apiFetch<AssistantConversationDetail>(
    `/api/v1/assistant/conversations/${id}`,
  );
}

export function deleteAssistantConversation(id: string): Promise<void> {
  return apiFetch<void>(`/api/v1/assistant/conversations/${id}`, {
    method: "DELETE",
  });
}
